### DISCUSSION TITLE
How to wrap a component?

### DISCUSSION THREAD

**Discussion by st-pasha** [flame-engine/flame#911] [Q&A] [answered]
Suppose I have a component `X` which is already mounted somewhere in a component tree: `A->B->...->V->X`. What I want is to wrap `X` into another component `W`, so that the component tree becomes `A->B->...->V->W(X)`. The component `W` should assume `X`'s identity: the priority and the place in the rendering sequence.

At first, I thought this would be as simple as `X.parent.replaceChild(X, W(X))`, but there is no `replaceChild()` function or anything remotely similar.

Do you guys have any suggestions for how to achieve this?

**Comment by spydon** [comment, marked as answer]
Edit: this was solved by #906 and `component.changeParent`

Good idea! This should be pretty easy to implement.
With `replaceChild` I would expect `W` to fully take `X` place in the tree though, and not wrap `X`, but maybe that is what you meant with the `W(X)` notation, that `X` is added to `W` there already?
I'm guessing we need quite a lot of optional parameters for the method too, since for a `PositionComponent` you might want it to take the `position`, `size`, `scale` or `angle` of the component that it replaces too.

Right now you would have to do something like this:
```dart
v.remove(x);
w.changePriorityWithoutResorting(x.priority);
// wait for next tick
x.shouldRemove = false;
w.addChild(x);
v.add(w);
```

After #906 you could do this instead:
```dart
w.changePriorityWithoutResorting(x.priority);
v.add(w);
x.changeParent(w);
```

So with this suggestion it would be something like this?
```dart
v.replaceChild(x, w, inheritPriority: true, inheritPosition: true ...);
w.add(x);
```

**Reply by st-pasha** [reply]
Well, I was thinking of a simple component like
```
class W extends Component {
  final Component _wrapped;
  W(this._wrapped);
  
  @override
  void update(double dt) { /* do something here */ }
  @override
  void render(Canvas canvas) { /* or maybe do something here */ }
}
```
Examples where this could be useful:
  - cast _slow_, or _freeze time_ on an NPC by passing a fraction of the real time `dt`;
  - modify `canvas` before drawing, i.e. apply a blend effect, or use a `Layer` as a component, or apply a canvas transform if for whatever reason the regular transform is not enough;
  - draw a special effect under or over a component, e.g. a visual for a spell effect;
  - render `X` and then post-process the image, such as adding blur or outline;
  - temporarily replace `X` with completely different visual, and then revert back after some time.

**Reply by st-pasha** [reply]
> After #906 you could do this instead:
> ```
> w.changePriorityWithoutResorting(x.priority);
> v.add(w);
> x.changeParent(w);```

There could be a problem if the components in the tree have the default priority. For example, `Root` has 3 children `[A, B, C]` all at the same default priority 0. If I now add `W` and reparent `A`, then root's children will be `[B, C, W->A]`, and so `W` did not take `A`'s place properly.

**Reply by spydon** [reply]
Alright, so we have two different things that are discussed here then:
 * `component.replaceChild`
 * A wrap component
 
Do you mean that `W` would live in Flame? If so then that could handle the transfer of the component to its own tree.

But I wonder if this solution is suitable, it seems like there would be a lot of tree modifications when a component is wrapped in many components that are added and removed. Maybe a new set, similar to `children` could be introduced which runs before the components own update/render functions? Not fully satisfied with such a solution either though, but it would be more efficient.

**Reply by spydon** [reply]
> > After #906 you could do this instead:
> > ```
> > w.changePriorityWithoutResorting(x.priority);
> > v.add(w);
> > x.changeParent(w);```
> > ```
> 
> There could be a problem if the components in the tree have the default priority. For example, `Root` has 3 children `[A, B, C]` all at the same default priority 0. If I now add `W` and reparent `A`, then root's children will be `[B, C, W->A]`, and so `W` did not take `A`'s place properly.

True, this is something that would need to be implemented in the `OrderedSet` then.

**Reply by st-pasha** [reply]
> Do you mean that W would live in Flame?

W is just any component. You could replace `X` with anything you want. Maybe a pawn becomes a Queen, or an enemy corpse turns into a Zombie. Some kinds of wrappers can be provided by Flame too, similar to Effects.
